test/versioned/kafkajs/kafka.tap.js

```diff
@@ -43,24 +43,182 @@ tap.afterEach(async (t) => {
   await t.context.producer.disconnect()
 })
 
-tap.test('stub', async (t) => {
-  const { consumer, producer, topic } = t.context
+tap.test('send records correctly', (t) => {
+  t.plan(4)
+
+  const { agent, consumer, producer, topic } = t.context
   const message = 'test message'
 
-  await consumer.subscribe({ topics: [topic], fromBeginning: true })
-  const testPromise = new Promise((resolve) => {
-    consumer.run({
-      eachMessage: async ({ message: actualMessage }) => {
-        t.equal(actualMessage.value.toString(), message)
-        resolve()
-      }
+  agent.on('transactionFinished', (tx) => {
+    const name = `MessageBroker/Kafka/Topic/Produce/Named/${topic}`
+    const segment = tx.agent.tracer.getSegment()
+
+    const foundSegment = segment.children.find((s) => s.name.endsWith(topic))
+    t.equal(foundSegment.name, name)
+
+    const metric = tx.metrics.getMetric(name)
+    t.equal(metric.callCount, 1)
+
+    t.end()
+  })
+
+  helper.runInTransaction(agent, async (tx) => {
+    await consumer.subscribe({ topic, fromBeginning: true })
+    const promise = new Promise((resolve) => {
+      consumer.run({
+        eachMessage: async ({ message: actualMessage }) => {
+          t.equal(actualMessage.value.toString(), message)
+          t.match(actualMessage.headers['x-foo'].toString(), 'foo')
+          resolve()
+        }
+      })
+    })
+    await utils.waitForConsumersToJoinGroup({ consumer })
+    await producer.send({
+      acks: 1,
+      topic,
+      messages: [
+        {
+          key: 'key',
+          value: message,
+          headers: {
+            'x-foo': 'foo'
+          }
+        }
+      ]
+    })
+    await promise
+
+    tx.end()
+  })
+})
+
+tap.test('send passes along DT headers', (t) => {
+  // The intent of this test is to verify the scenario:
+  //
+  // 1. A service receives a request
+  // 2. The service builds a payload for Kafka
+  // 3. The produced Kafka data includes the distributed trace data that was
+  // provided to the service handling the request.
+
+  t.plan(10)
+
+  const now = Date.now
+  Date.now = () => 1717426365982
+  t.teardown(() => {
+    Date.now = now
+  })
+
+  const { agent, consumer, producer, topic } = t.context
+  const messages = ['one', 'two', 'three']
+
+  // The traceparent, tracestate, and agent.config lines are utilized to
+  // simulate the inbound distributed trace that we are trying to validate.
+  const traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00'
+  const tracestate = '42@nr=0-0-account_1-app_1-span_1-txn_1-0-0.123000-' + Date.now()
+  agent.config.account_id = 'account_1'
+  agent.config.primary_application_id = 'app_1'
+  agent.config.trusted_account_key = 42
+
+  agent.on('transactionFinished', (tx) => {
+    t.equal(tx.isDistributedTrace, true)
+
+    const headers = {}
+    tx.traceContext.addTraceContextHeaders(headers)
```

Comment:
instead of doing this you could assert these values in the `eachMessage` handler. they are buffers obviously so you'd have to do `const tracestate = actualMessage.headers.tracestate.toString()`

Reply:
The traceparent will be something generated from within asn it won't be the traceparent that was accepted but the first part should be the same `42@nr=0-0-account_1-app_1`